# Credit Risk Analysis — Preprocesamiento
## Objetivo
Preparar el dataset limpio del EDA para entrenar el modelo de regresión logística.
Incluye: selección de variables, escalado y manejo de clase desbalanceada.

In [1]:
#Importamos librerias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme(style='whitegrid')

In [2]:
df= pd.read_csv('../data/cs-training-limpio.csv')
print(df.shape)
print(df.isnull().sum())

(149986, 11)
SeriousDlqin2yrs                        0
RevolvingUtilizationOfUnsecuredLines    0
age                                     0
NumberOfTime30-59DaysPastDueNotWorse    0
DebtRatio                               0
MonthlyIncome                           0
NumberOfOpenCreditLinesAndLoans         0
NumberOfTimes90DaysLate                 0
NumberRealEstateLoansOrLines            0
NumberOfTime60-89DaysPastDueNotWorse    0
NumberOfDependents                      0
dtype: int64


In [3]:
#variable objetivo
y=df['SeriousDlqin2yrs']

#variable predictoras - eliminamos las redundantes o las que no vamos a usar
X= df.drop(columns=['SeriousDlqin2yrs','NumberOfTime30-59DaysPastDueNotWorse','NumberOfTime60-89DaysPastDueNotWorse'])

print('X: ', X.shape)
print('y: ', y.shape)


X:  (149986, 8)
y:  (149986,)


In [4]:
from sklearn.model_selection import train_test_split

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print('X_train:', X_train.shape)
print('X_test:', X_test.shape)

X_train: (119988, 8)
X_test: (29998, 8)


In [6]:
from sklearn.preprocessing import StandardScaler

scaler= StandardScaler()

X_train_scaled= scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [7]:
from imblearn.over_sampling import SMOTE

In [8]:
smote = SMOTE(random_state=42)
X_train_balanced, y_train_balanced = smote.fit_resample(X_train_scaled, y_train)

In [9]:
print('Distribución antes del SMOTE:')
print(y_train.value_counts())
print('\nDistribución después del SMOTE:')
print(pd.Series(y_train_balanced).value_counts())

Distribución antes del SMOTE:
SeriousDlqin2yrs
0    111938
1      8050
Name: count, dtype: int64

Distribución después del SMOTE:
SeriousDlqin2yrs
0    111938
1    111938
Name: count, dtype: int64


## Preprocesamiento completado

1. Variables seleccionadas: 8 predictoras + 1 objetivo
   - Eliminadas por multicolinealidad: NumberOfTime30-59Days y NumberOfTime60-89Days
2. División: 80% entrenamiento (119,988) / 20% prueba (29,998)
3. Escalado: StandardScaler aplicado a X_train y X_test
4. Balance de clases con SMOTE:
   - Antes: 111,938 vs 8,050
   - Después: 111,938 vs 111,938
   - SMOTE aplicado solo al entrenamiento para preservar realismo en la evaluación